### ISSUE TITLE
Sporadic Websocket Timeouts

### ISSUE THREAD

**Issue by connor4312** [balderdashy/sails#2768]
We're running Sails 0.11 in a clustered system with sticky sessions, and have a client attempted to connect over websockets. He is able to connect, however it appears that Sails times out the socket for him.

![](http://puu.sh/gJy5e/d1f7c6fed6.png)

Notice the lack of heartbeat responses. After a time, his socket.io client thinks the connection is dead and attempts to reconnect him (with the same results).

The socket connection is being load balanced and proxied through nginx, which records the following error:

```
2015/03/21 16:23:58 [info] 13919#0: *18553690 upstream timed out (110: Connection timed out) while proxying upgraded connection, client: xxx.xxx.xxx.xxx, server: xxx.xxx, request: "GET /socket.io/?__sails_io_sdk_version=0.11.0&__sails_io_sdk_platform=node&__sails_io_sdk_language=javascript&EIO=3&transport=websocket HTTP/1.1", upstream: "xxx/sails.sock.24:/socket.io/?__sails_io_sdk_version=0.11.0&__sails_io_sdk_platform=node&__sails_io_sdk_language=javascript&EIO=3&transport=websocket", host: "xxx.xxx"
```

From this log (upstream timeout) it appears that the application server is the cause. Our sockets config uses default settings, we don't use socket.io-redis or related adapters. We do not register lifecycle callbacks for the socket, and the only particular manipulation we do on the websocket itself are in some endpoints which are not touched in the above cases. The only global shimming we do is via NewRelic. We have attempted removing that, but the issue persists.

Grepping our logs, this issue is specific to a handful of IPs, making up about 0.2% of our total traffic. For those clients, the issue _always_ occurs whenever they attempt to load the site. The clients do not share a common service provider, location, network, or even continent.

He has tried this with modern Chrome, Firefox, and Internet Exploder. It persists for him no matter what browser he uses, including using "incognito" mode, but only on one of his computers (his other computers also share the same IP).

As far as we can tell, this issue is a result of some internal handling within Sails.

**Comment by crh3675** [comment]
Interesting, one client eh? Sounds like a shady computer with potential malware interrupting socket requests.

**Comment by connor4312** [comment]
Multiple clients several different IP addresses. If malware on the client-side can cause the socket.io _server_ to time out, as nginx reports, that's definitely a concern.

**Comment by connor4312** [comment]
We had him run a scan for malware with MBAM, but that did nothing. For this particular client, when he installed Windows Updates, it ran. (He had rebooted his computer previously, so that was not the cause). 

So it's resolved in this particular case, but it's a bit disturbing that unknown client issues can consistently cause upstream timeouts on the server.

**Comment by rrupesh** [comment]
I've been recently having the same issues with sails socket. I was using mongo as db and also for sessions. I then shifted to mysql for db and mongo for sessions and everything seems to be working fine. no more issues with sockets or http requests.

**Comment by mikermcneil** [comment]
Thanks for posting, @connor4312.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
